util/flocks/src/frwlock/mod.rs

```diff
@@ -0,0 +1,200 @@
+pub mod read_guard;
+pub mod write_guard;
+
+pub use write_guard::FrwLockWriteGuard;
+pub use read_guard::FrwLockReadGuard;
+use rustix::{
+    fs::{flock, FlockOperation},
+    fd::AsFd,
+};
+use std::cell::UnsafeCell;
+use thiserror::Error;
+use tokio::task::yield_now;
+
+#[derive(Debug, Error)]
+pub enum FrwLockError {
+    #[error("Lock is not available")]
+    LockNotAvailable,
+    #[error("File error: {0}")]
+    FileError(#[from] std::io::Error),
+}
+
+/// A file-based read-write lock.
+/// This only mutually excludes processes trying to violate the lock, not the same process--which is not considered contention.
+/// If you want to prevent contention within the same process, you should wrap this in your preferred synchronization primitive.
+pub struct FrwLock<T: AsFd> {
+    cell: UnsafeCell<T>,
+}
+
+impl<T: AsFd> FrwLock<T> {
+    pub fn new(file: T) -> Self {
+        Self {
+            cell: UnsafeCell::new(file),
+        }
+    }
+
+    pub(crate) fn try_write(&self) -> Result<FrwLockWriteGuard<T>, FrwLockError> {
+        let file = unsafe { &*self.cell.get() };
+        match flock(file, FlockOperation::NonBlockingLockExclusive) {
```

Comment:
`match file_lock`

Reply:
I didn't name this one flock. It comes from `rustix` which is wrapping Unix APIs: https://linux.die.net/man/2/flock

Reply:
I wasn't aware of the term, in that case I don't mind, if we go with this implementation.